[6.5][NET] Added application service description (#1494)
Samsung/tizen-docs: docs/application/dotnet/guides/applications/service_application.md

The service application type allows creating reusable and independent parts which are really important in bigger projects where we can easily split responsibilities of the application into different parts. 
Consider the speedometer application designed for average speed measurement; the service part, in this case, is responsible for reading speed from the device GPS module, it also calculates the average speed. The UI communicates with the service when required and shows the measured values to the user. 
This approach allows reusing modules between different applications such as cyclist applications, automotive solutions, and sport activity apps. The core module, with business logic, works in the background and performs the task it was designed for until it gets closed by system events like `OnLowMemory` or `OnLowBattery`. The activity of the core module remains unaffected by any user action such as accidental closing of the app UI.

> The main service application API features include: 

Reviewer: API names are common compound nouns. So capitalized one is correct.
Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/naming-rules.md#proper-nouns
```suggestion
The main Service Application API features include: 
```

# Service Applications 

Service applications are Tizen .NET applications with no graphical user interface that run in the background. They can be very useful in performing activities (such as getting sensor data in the background) that need to run periodically or continuously, but do not require any user intervention.

> The service application type allows creating reusable and independent parts which are really important in bigger projects where we can easily split responsibilities of the application into different parts. As an example, consider the speedometer application designed for average speed measurement. The service part in this case is responsible for reading speed from the device GPS module, it also calculates average speed. The UI communicates with the service when it is visible and shows measured values. This approach allows reusing modules between different applications such as cyclist applications, automotive solutions, and sport activity apps. The core module with business logic works in the background and carry out designed things until it gets closed by the system events like `OnLowMemory` or `OnLowBattery` where the UI app could be accidentally closed by the user.

Author: What I mean is that the user can accidentally close the application while collecting the needed data, for example by launching another application or pressing the back button. If the application UI doesn't handle such cases in an appropriate way, e.g. the data collection process may be closed. If the application is divided into service such case can be handled much easier without losing functionality. 

Translated with www.DeepL.com/Translator (free version)

> The service application type allows creating reusable and independent parts which are really important in bigger projects where we can easily split responsibilities of the application into different parts. As an example, consider the speedometer application designed for average speed measurement. The service part in this case is responsible for reading speed from the device GPS module, it also calculates average speed. The UI communicates with the service when it is visible and shows measured values. This approach allows reusing modules between different applications such as cyclist applications, automotive solutions, and sport activity apps. The core module with business logic works in the background and carry out designed things until it gets closed by the system events like `OnLowMemory` or `OnLowBattery` where the UI app could be accidentally closed by the user.

Reviewer: I've made some significant changes to this paragraph; breaking it down to smaller, easily understandable sentences. If the suggestion is acceptable to you, please incorporate it.
```suggestion
The service application type allows creating reusable and independent parts which are really important in bigger projects where we can easily split responsibilities of the application into different parts. 
Consider the speedometer application designed for average speed measurement; the service part, in this case, is responsible for reading speed from the device GPS module, it also calculates the average speed. The UI communicates with the service when required and shows the measured values to the user. 
This approach allows reusing modules between different applications such as cyclist applications, automotive solutions, and sport activity apps. The core module, with business logic, works in the background and performs the task it was designed for until it gets closed by system events like `OnLowMemory` or `OnLowBattery`. The activity of the core module remains unaffected by any user action such as accidental closing of the app UI.
```

```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns="http://tizen.org/ns/packages" api-version="2.4" package="org.tizen.test" version="1.0.0">
   <ui-application appid="org.tizen.test" exec="text" type="capp" multiple="false" taskmanage="true" nodisplay="false">
      <icon>rest.png</icon>
      <label>rest</label>
      <!--For API version 2.4 and higher-->
      <background-category value="media"/>
      <background-category value="download"/>
      <background-category value="background-network"/>
   </ui-application>
   <service-application appid="org.tizen.test-service" exec="test-service" multiple="false" type="capp">
      <background-category value="background-network"/>
      <background-category value="location"/>
   </service-application>
</manifest>
```

## Manage the service application

The following code snippets show the service application backbone generated from the Tizen templates.
The service template creates callback stubs to fill. Service application type is defined in the `Tizen.Applications` namespace, so it should be included at the top of the file. 

```csharp
using Tizen.Applications;
```

> The `ServiceApplication` has no UI, so `OnPause()` and `OnResume()` are not defined. Besides of that it defines similar callbacks to other app types like `OnCreate()` and `OnTerminate()` lifecycle callbacks, and `OnLowBattery()`, `OnLowMemory()`, `OnLocaleChanged()` system callbacks.

Reviewer: Line 93: Please check this, should this be "OnAppTerminate()" or "OnTerminate()"?

Author: OnTerminate() -> I changed the image

`OnCreate()` method is used to take necessary actions, before the main event loop starts. Place the initialization code here:

```csharp
        protected override void OnCreate()
        {
            base.OnCreate();
            
            //Initialization
        }
```

`OnAppConrolReceived()` callback is the most important callback since it is responsible for data exchange between the service app and other applications:

```csharp
        protected override void OnAppControlReceived(AppControlReceivedEventArgs e)
        {
            base.OnAppControlReceived(e);
        }
```

> `OnTerminate()` callback is used to take necessary actions when the application is terminating. It releases all resources, especially any allocations and shared resources.

Reviewer: Line 115: same here, should this be "OnAppTerminate()" or "OnTerminate()"?

```csharp

>         protected override void OnTerminate()

Reviewer: Line 118: should this be "OnAppTerminate()" or "OnTerminate()"?

Author: OnTerminate() -> I changed the image

        {

>             base.OnTerminate();

Reviewer: Line 120: should this be "OnAppTerminate()" or "OnTerminate()"?

            app.Run(args);
        }
    }
}
```

## Application attributes
Describe your service application attributes in the manifest file. The attributes determine the application behavior. The following code example illustrates how you can define the attributes:

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns="http://tizen.org/ns/packages" api-version="6" package="org.tizen.example.servicesample" version="1.0.0">
  <profile name="common" />
  <service-application appid="org.tizen.example.servicesample"
					exec="servicesample.dll"
					type="dotnet"
					multiple="false"
					taskmanage="false"
					nodisplay="true"
                    api-version="8">
    <label>servicesample</label>
    <icon>servicesample.png</icon>
  </service-application>
</manifest>
```

- `auto-restart`

  If set to `true`, the application restarts whenever it terminates abnormally. If the application is running, it is launched after installing or updating the package.

- `on-boot`

  If set to `true`, the application launches on boot time, and after installing or updating the package. The application does not start, if this attribute is removed after updating the package.

> 

Author: Of course you're right. I removed on-boot and auto-restart information, thanks!

The service application can be explicitly launched by a UI application. They can also be launched conditionally.

The list of running service applications is available in the task switcher system application; however, no events occur if the user selects a service application from the task switcher. The main menu does not contain icons for service applications. Multiple service applications can be running simultaneously with other service and UI applications.

## Application states

The following figure and table describe the service application states.

**Figure: Service application lifecycle**
![Service Lifecycle](media/service_application_states.png)

**Table: Service application states**

| State        | Description                         |
|--------------|-------------------------------------|
| `READY`      | Application is launched.            |
| `CREATED`    | Application starts the main loop.   |
| `RUNNING`    | Application runs in the background. |
| `TERMINATED` | Application is terminated.          |

> Since a service application has no UI, it does not have a pause state. Since Tizen 2.4, the service application can go into the suspended state. Basically, the service application is running in the background by its nature; so the platform does not allow running the service application unless the application has a background category defined in its manifest file. However, when the UI application that is packaged with the service application is running on the foreground, the service application is also regarded as a foreground application and it can be run without a designated background category.

Reviewer: ```suggestion
Since a service application has no UI, it does not have a pause state. Since Tizen 2.4, the service application can go into the suspended state. Basically, the service application is running in the background by its nature; so the platform does not allow running the service application unless the application has a background category defined in its manifest file. However, when the UI application that is packaged with the service application is running in the foreground, the service application is also regarded as a foreground application and it can be run without a designated background category.
```

### Describe background category

An application with a background running capability must declare the background category in its manifest file:

```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns="http://tizen.org/ns/packages" api-version="2.4" package="org.tizen.test" version="1.0.0">
   <ui-application appid="org.tizen.test" exec="text" type="capp" multiple="false" taskmanage="true" nodisplay="false">
      <icon>rest.png</icon>
      <label>rest</label>
      <!--For API version 2.4 and higher-->
      <background-category value="media"/>
      <background-category value="download"/>
      <background-category value="background-network"/>
   </ui-application>
   <service-application appid="org.tizen.test-service" exec="test-service" multiple="false" type="capp">
      <background-category value="background-network"/>
      <background-category value="location"/>
   </service-application>
</manifest>
```

## Manage the service application

The following code snippets show the service application backbone generated from the Tizen templates.

> The service template creates callback stubs to fill. Service application type is defined in the `Tizen.Applications` namespace, so it should be included at the top of the file. 

Reviewer: Code snippet should be preceded by a colon.
```suggestion
The service template creates callback stubs to fill. The service application type is defined in the `Tizen.Applications` namespace, so it should be included at the top of the file:
```

```csharp
using Tizen.Applications;
```

The `ServiceApplication` has no UI, so `OnPause()` and `OnResume()` are not defined. Besides of that it defines similar callbacks to other app types like `OnCreate()` and `OnTerminate()` lifecycle callbacks, and `OnLowBattery()`, `OnLowMemory()`, `OnLocaleChanged()` system callbacks.

`OnCreate()` method is used to take necessary actions, before the main event loop starts. Place the initialization code here:

```csharp
        protected override void OnCreate()
        {
            base.OnCreate();
            
            //Initialization
        }
```

> `OnAppConrolReceived()` callback is the most important callback since it is responsible for data exchange between the service app and other applications:

Reviewer: ```suggestion
`OnAppControlReceived()` callback is the most important callback since it is responsible for data exchange between the service app and other applications:
```

```csharp
        protected override void OnAppControlReceived(AppControlReceivedEventArgs e)
        {
            base.OnAppControlReceived(e);
        }
```

> `OnTerminate()` callback is used to take necessary actions when the application is terminating. It releases all resources, especially any allocations and shared resources.

Reviewer: Code snippet should be preceded by a colon.
```suggestion
`OnTerminate()` callback is used to take necessary actions when the application is terminating. It releases all resources, especially any allocations and shared resources:
```

**Table: Attribute combinations**

| `auto-restart` | `on-boot` | After normal termination   | On forced close            | On Reboot                           | After package installation | After package update       |
|----------------|-----------|----------------------------|----------------------------|-------------------------------------|----------------------------|----------------------------|
| `FALSE`        | `FALSE`   | Not launched automatically | Not launched automatically | Not launched after reboot           | Not launched               | Not launched automatically |
| `FALSE`        | `TRUE`    | Not launched automatically | Not launched automatically | Launched automatically after reboot | Launched                   | Launched automatically     |
| `TRUE`         | `FALSE`   | Launched automatically     | Launched automatically     | Not launched after reboot           | Not launched               | Launched automatically     |
| `TRUE`         | `TRUE`    | Launched automatically     | Launched automatically     | Launched automatically after reboot | Launched                   | Launched automatically     |

## Related information

>   * Dependencies

Reviewer: Although the "*" appears as a bullet point in the staging area, requesting to use a dash "-" here for consistencies sake. 
```suggestion
  - Dependencies
```

The service application type allows to create reusable and independent parts which is really important in bigger projects where we can easily split responsibilities of the application to different parts. As an example, consider the speedometer application designed for average speed measurement. Service part in this case is responsible for reading speed from the device GPS module. It also calculates average speed. The UI communicates with the service when it is visible and shows measured values. That approach allows to reuse module between different applications such as a cyclist application, automotive solutions, sport activity apps. Core module with business logic works in the background and carry out designed things until it gets closed by the system evens like `OnLowMemory` or `OnLowBattery` where UI app could be accidentally closed by the user.

The main Service Application API features include: 
-  Application States:

>     A Tizen native service application has different states which it transitions through during its life-cycle.

Reviewer: This page seems to be copied from the page under native so there are several 'native' strings. Please incorporate the suggestion if you think this is correct or let me know otherwise.
```suggestion
    A Tizen service application has different states which it transitions through during its life-cycle.
```

- Event Callbacks:
  

>     The service application can receive both basic system events and application state change events. You can register handlers for these events to react them.

Author: Maybe response to them will be better? What do you think ?

> [!NOTE]
> Since Tizen 4.0, even if the background network category is declared, the running application stops if the network is not connected.

### Describing background category

An application with a background running capability must declare the background category in its manifest file:

```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns="http://tizen.org/ns/packages" api-version="2.4" package="org.tizen.test" version="1.0.0">
   <ui-application appid="org.tizen.test" exec="text" type="capp" multiple="false" taskmanage="true" nodisplay="false">
      <icon>rest.png</icon>
      <label>rest</label>
      <!--For API version 2.4 and higher-->
      <background-category value="media"/>
      <background-category value="download"/>
      <background-category value="background-network"/>
   </ui-application>
   <service-application appid="org.tizen.test-service" exec="test-service" multiple="false" type="capp">
      <background-category value="background-network"/>
      <background-category value="location"/>
   </service-application>
</manifest>
```

> > [!NOTE]

Reviewer: Note in line 84: These are about Tizen version before Tizen 2.4, but Tizen .NET started from Tizen 4.0, so I'm not sure if these notes are used to .NET aswell. Please consider revising this.

> <manifest xmlns="http://tizen.org/ns/packages" api-version="2.3" package="org.tizen.test" version="1.0.0">
>    <ui-application appid="org.tizen.test" exec="text" type="capp" multiple="false" taskmanage="true" nodisplay="false">
>       <icon>rest.png</icon>
>       <label>rest</label>
>       <!--For API version lower than 2.4-->
>       <metadata key="http://tizen.org/metadata/background-category/media"/>
>       <metadata key="http://tizen.org/metadata/background-category/download"/>
>       <metadata key="http://tizen.org/metadata/background-category/background-network"/>
>    </ui-application>
>    <service-application appid="org.tizen.test-service" exec="test-service" multiple="false" type="capp">
>       <metadata key="http://tizen.org/metadata/background-category/background-network"/>
>       <metadata key="http://tizen.org/metadata/background-category/location"/>
>    </service-application>
> </manifest>
> ```
>
> The `<metadata key="http://tizen.org/metadata/bacgkround-category/<value>"/>` element has no effect on Tizen 2.3 devices, but on Tizen 2.4 and higher devices, it has the same effect as the `<background-category>` element.

> ## Managing the service application 

Reviewer: Gerund case, please consider revising it.
```suggestion
## Manage the service application 
```

Following code snippets shows the service application backbone generated from the Tizen templates.
The service template creates callback stubs to fill. Service application type is defined in the `Tizen.Applications` namespace, so it should be included at the top of the file. 

```csharp
using Tizen.Applications;
```

> The `ServiceApplication` has no UI, so `OnPause()` and `OnResume()` are not defined. Besides of that it defines similar callbacks to other app types like `OnCreate()` and `OnTerminate()` lifecycle callbacks, and `OnLowBattery()`, `OnLowMemory()`, `OnLocaleChanged()` system callbacks.

Reviewer: In line 29, in the image, OnAppTerminate() is used to terminate, but line 115 says, OnTerminate() is used.
Please fix this.

Reviewer: > In line 29, in the image, OnAppTerminate() is used to terminate, but line 115 says, OnTerminate() is used. Please fix this.

Line 93: Please check this, my previous concern still remains.

        {
            base.OnRegionFormatChanged(e);
        }
```

```csharp
        static void Main(string[] args)
        {
            App app = new App();
            app.Run(args);
        }
    }
}
```

## Application attributes
Describe your service application attributes in the manifest file. The attributes determine the application behavior. The following code example illustrates how you can define the attributes:

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns="http://tizen.org/ns/packages" api-version="6" package="org.tizen.example.servicesample" version="1.0.0">
  <profile name="common" />
  <service-application appid="org.tizen.example.servicesample"
					exec="servicesample.dll"
					type="dotnet"
					multiple="false"
					taskmanage="false"
					nodisplay="true"
                    api-version="8">
    <label>servicesample</label>
    <icon>servicesample.png</icon>
  </service-application>
</manifest>
```

- `auto-restart`

  If set to `true`, the application restarts whenever it terminates abnormally. If the application is running, it is launched after installing or updating the package.

  >[!NOTE]

>   > This attribute is not supported on Tizen wearable devices. Since Tizen 2.4, this attribute is not supported on all Tizen devices. Because of this, the `auto-restart` attribute used in a lower API version package than 2.4 is ignored on devices with the Tizen platform version 2.4 and higher.

Reviewer: Note in line 216: Same issue here, these are about Tizen version before Tizen 2.4, but Tizen .NET started from Tizen 4.0, so I'm not sure if these notes are used to .NET as well. Please consider revising this.

- `on-boot`

  If set to `true`, the application launches on boot time, and after installing or updating the package. The application does not start, if this attribute is removed after updating the package.

  >[!NOTE]

>   > This attribute is not supported on Tizen wearable devices. Since Tizen 2.4, this attribute is not supported on all Tizen devices. Because of this, the `on-boot` attribute used in a lower API version package than 2.4 is ignored on devices with the Tizen platform version 2.4 and higher.

Reviewer: Note in line 223: Same issue here, these are about Tizen version before Tizen 2.4, but Tizen .NET started from Tizen 4.0, so I'm not sure if these notes are used to .NET as well. Please consider revising this.

>   > This attribute is not supported on Tizen wearable devices. Since Tizen 2.4, this attribute is not supported on all Tizen devices. Because of this, the `on-boot` attribute used in a lower API version package than 2.4 is ignored on devices with the Tizen platform version 2.4 and higher.

Author: I based on native api documentation. We cannot use C# API in lower than 4.5 so I removed this note.

> # Service Applications 

Reviewer 2: Since this is a new topic, it need to be included in the overview topic :(https://docs1.stg.tizen.org/staging/1494/application/dotnet/guides/applications/overview/)

Kindly check and update overview topic and include in the PR.
